File: oss/client-mobile/android/src/main/java/expo/modules/omniauth/OmniAuthModule.kt
package expo.modules.omniauth

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import com.omniauth.core.Vault

class OmniAuthModule : Module() {
  private var activeVault: Vault? = null

  override fun definition() = ModuleDefinition {
    Name("OmniAuth")

    AsyncFunction("createVault") { password: String ->
      try {
        val vault = Vault(password)
        activeVault = vault
        return@AsyncFunction "SUCCESS"
      } catch (e: Exception) {
        throw Exception("Vault Creation Failed: ${e.message}")
      }
    }

    AsyncFunction("getPublicKey") { password: String ->
      val vault = activeVault ?: throw Exception("No active vault")
      try {
        return@AsyncFunction vault.getPublicSigningKey(password)
      } catch (e: Exception) {
        throw Exception("Failed to get public key: ${e.message}")
      }
    }

    AsyncFunction("signChallenge") { password: String, message: String ->
      val vault = activeVault ?: throw Exception("No active vault")
      try {
        return@AsyncFunction vault.sign(password, message)
      } catch (e: Exception) {
        throw Exception("Signing failed: ${e.message}")
      }
    }

    Function("exportBlob") {
      val vault = activeVault ?: throw Exception("No active vault")
      return@Function vault.exportEncryptedBlob()
    }

    AsyncFunction("restoreVault") { encryptedBlob: String ->
      try {
        activeVault = Vault.newWithBlob(encryptedBlob)
        return@AsyncFunction "SUCCESS"
      } catch (e: Exception) {
        throw Exception("Vault Restore Failed: ${e.message}")
      }
    }

    Function("lock") {
      activeVault = null
    }
  }
}

File: oss/client-mobile/ios/OmniAuthModule.swift
import ExpoModulesCore
import omniauth_core

public class OmniAuthModule: Module {
  private var activeVault: Vault? = nil
  private var activePassword: String? = nil

  public func definition() -> ModuleDefinition {
    Name("OmniAuth")

    AsyncFunction("createVault") { (password: String) -> String in
      do {
        let vault = try Vault(masterPassword: password)
        self.activeVault = vault
        self.activePassword = password
        return "SUCCESS"
      } catch {
        throw Exception(name: "VaultCreationError", description: error.localizedDescription)
      }
    }

    AsyncFunction("getPublicKey") { (password: String) -> String in
      guard let vault = self.activeVault else {
        throw Exception(name: "NoVault", description: "No active vault")
      }
      do {
        return try vault.getPublicSigningKey(password: password)
      } catch {
        throw Exception(name: "GetKeyError", description: error.localizedDescription)
      }
    }

    AsyncFunction("signChallenge") { (password: String, message: String) -> String in
      guard let vault = self.activeVault else {
        throw Exception(name: "NoVault", description: "No active vault")
      }
      do {
        return try vault.sign(password: password, message: message)
      } catch {
        throw Exception(name: "SignError", description: error.localizedDescription)
      }
    }

    Function("exportBlob") { () -> String in
      guard let vault = self.activeVault else {
        throw Exception(name: "NoVault", description: "No active vault")
      }
      do {
        return try vault.exportEncryptedBlob()
      } catch {
        throw Exception(name: "ExportError", description: error.localizedDescription)
      }
    }

    AsyncFunction("restoreVault") { (encryptedBlob: String) -> String in
      do {
        self.activeVault = try Vault.newWithBlob(encryptedBlobStr: encryptedBlob)
        return "SUCCESS"
      } catch {
        throw Exception(name: "RestoreError", description: error.localizedDescription)
      }
    }

    Function("lock") {
      self.activeVault = nil
      self.activePassword = nil
    }
  }
}
